include/exec/task.hpp

```diff
@@ -257,6 +358,29 @@ namespace exec {
           this->__data_.template emplace<2>(std::current_exception());
         }
 
+        template <class _Awaitable>
+        decltype(auto) await_transform(_Awaitable&& __awaitable) noexcept
+          requires(
+            __completion_scheduler_provider<_Context>
+            && !stdexec::__decays_to<_Awaitable, __stick_on_scheduler_t>)
+        {
+          return stdexec::as_awaitable(
+            stdexec::transfer(
+              (_Awaitable&&) __awaitable,
+              stdexec::get_completion_scheduler<stdexec::set_value_t>(__context_)),
```

Comment:
Rather than relying on the sender's completion scheduler, the `co_await` should squirrel the current scheduler away in a custom awaitable , and then post back to that scheduler on resume.

Reply:
Ah, okay.

Reply:
BTW, I particularly appreciate the greenfield effort you're making here. I implemented this feature in libunifex and feel I did a poor job of it. Don't emulate it, at least not blindly.

Reply:
I have taken a look into your PR in libunifex. After I have understood all the whys I will happly apply all the changes

Reply:
I don't think registering a cleanup action is necessary, FWIW. That's something I'd like to try to move away from.

Reply:
I still try to find the goals/requirements of my "task" (not `exec::task`). Sorry I should have asked that way before.

I've pondered over your implementation and the difference to mine. Is the goal, that each `task` object has an associated `completion_scheduler<set_value_t>` which it is guaranteed to complete on? Then my implementation fails to satisfy that requirement, because it only relies on the _current_ scheduler within the context, which might lead to a different scheduler in the end. But how am I supposed to start a task with a specified `scheduler`? I can imagine a promise constructor, which checks for a `scheduler_arg_t` argument analogue to `allocator_arg_t` elsewhere in the standard. But how was that in your original design?

Reply:
The goal is for the task to complete on whichever was the current scheduler when the task was started. So in general, a task will not know its completion scheduler.

To start a task on a particular scheduler, the idea is to just use the `on` algorithm, which sets the current scheduler. In other words, that part is already implemented.

Reply:
Is it correct to reschedule on the initial scheduler within task awaitable type? What if my task is used with other sender algorithms? I will need to add tests. I've played around with rescheduling in the final suspend. That yields good results on clang but ICE on gcc ...

Reply:
> Is it correct to reschedule on the initial scheduler within task awaitable type? What if task is used with other sender algorithms? I will need to add tests.

Reply:
I've found the problem and improved the implementation. Next, I will look into clang-12 and why its not working there.

Reply:
Regarding the clang-12 issue: The problem seems to be, that std::this_thread::get_id() reports a wrong thread id. When I step through the debugger in Release mode with debug symbols, I can see that the coro was scheduled on the correct thread but the ID remains unchanged. This happens only in Release mode and does not happen with clang-16. Can someone verify that?